---
title: "Methods of Differential Gene Expression Analysis"
author: "Lisa Hoeg, 1151916"
date: "BINF6210 A5: due Dec. 11, 2020"
output: pdf_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE, tidy.opts = list(width.cutoff=60))
```

## 1. Introduction

"Paragraph 1: What is the overarching topic you are interested in? Why is this scientifically important and interesting and/or societally relevant?"

-Choosing Option 7, exploring the methods of analyzing differential gene expression
-There are many competing popular platforms that use different statistical methods, each seem to have their most appropriate situations, no agreement on best
-articles to reference: Wang et al (2019; comparative analysis, not read yet, add if time); Tang et al (2015, evaluation of methods); De Paepe (2015, comparison of methods)
-could cite 1-2 of these, and the others in the conclusion to see how my results compare to existing literature

"Paragraph 2: You can outline what is an important sub-area of research or a gap in knowledge. (As an example, perhaps your broader theme is gene expression analysis, and in paragraph 2 you could narrow in to introduce the importance of statistical choices.)"

-Might focus on normalization methods and visualizations
-Starting with count data, not raw sequence reads
-Li et al(2020, library size normalization)
- a few more not yet read: Maza (2016, TMM/edgeR vs RLE/DESeq2 vs MRN);Hafemeister & Satija (2019, normalization and VS w NB regression)
-This paper is useful about visualizations McDermaid et al (2018), 

"Paragraph 3: What is the specific objective of your study? What kind of study are you performing? What will you do?"

Probably I should say something here.


## 2. Description of Data Set

-From Stilling et al. the original study had three groups of mice, conventional (CON), germ-free (GF), and ex-GF (born GF, colonized at weaning). They used different mice for the control group (naive) and the group subjected to a social interaction task. This makes 6 groups for the original study, I will subset and look at CON-SI/GF-SI, if time might add CON/GF naive (P). Including biological replicates, they had 40 samples, I will use 4 biological replicates of each, for a total of 8 samples. Count data was downloaded from GEO on 11/24, results matrix was downloaded on 11/27.


## 3. Code Section 1 – Data Acquisition, Exploration, Filtering, and Quality Control
Note: only coding of interest to the assignment and their outputs will be displayed in this section. A number of simple commands, such as loading libraries and "sanity checks" like verifying the class of an object will make use of the "include = FALSE" option to keep the output document easy to read. The full code will be present at the end of the PDF for grading purposes.

Preliminary Set-up & Data Acquisition

``` {r libraries, results = "hide"}
# All libraries were loaded at the beginning of the script for efficient organization.
#install.packages("stringr")
library(stringr)
#BiocManager::install("org.Mm.eg.db")
library(org.Mm.eg.db)
#BiocManager::install("edgeR")
library(edgeR)
#BiocManager::install("DESeq2")
library(DESeq2)
#install.packages("ggplot2")
library(ggplot2)
#install.packages("gridExtra")
library(gridExtra)
#install.packages("readxl")
library(readxl)
#BiocManager::install("vidger")
library(vidger)
```

```{r set-dir, results = "hide"}
# A variable, "dir" is used to define the working directory
# Keeping this here for now, before submitting delete this comment and change to empty string: ""
# User can specify directory here for downloading and saving files, or leave empty
dir <- "/Users/lisa/Documents/Bioinformatics/6210/A5/scripts_data/a5_data/"
```

```{r load-data}
# Download file with these two commands
# Count_URL <- paste("http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE114702", "format=file",
#"file=GSE114702_HTSeq_gene_counts.txt.gz", sep="&")
# download.file(Count_URL, paste0(dir, "GSE114702_HTSeq_gene_counts.txt.gz"))

# Read in file; first column is the gene identifiers, which will be used for row names
countdata_mice <- read.delim(paste0(dir, "GSE114702_HTSeq_gene_counts.txt.gz"), row.names = 1)
```

```{r load-checks, include = FALSE}
# Check data loaded as expected:
class(countdata_mice)
colnames(countdata_mice)
head(rownames(countdata_mice))
countdata_mice[1:5, c(1,6,18,30)]
```

```{r replicate-sample}
# In order to efficiently sample the biological replicates of the groups I wish to work with, I will first define a function to sample a single group at a time, and a companion function to sample all groups at once.
replicate_sample <- function(samples, group, size = 4) {
  gp_index <- grep(paste0("^", group), samples)
  sample_index <- sample(gp_index, size)
  return(sample_index)
}

replicate_sample_all <- function(samples, size = 4) {
  gps <- unique(str_remove(samples, "[0-9]+$"))
  full_index <- lapply(gps, FUN = function(x) replicate_sample(samples, x))
  full_index <- unlist(full_index)
  return(full_index)
}
```

```{r subset-data}
# With the function for sampling the biological replicates defined, I will extract 4 biological replicates for the social interaction group mice of the CON and GF types (CON.S & GF.S)
sample_names <- colnames(countdata_mice)
# Setting seed for reproducibility
set.seed(40404)
# Applying function for 2 groups of interest
col_S_CON_GF <- lapply(c("CON.S", "GF.S"), FUN = function(x) replicate_sample(sample_names, x))
col_S_CON_GF <- unlist(col_S_CON_GF)
# Subsetting data
counts_S_CON_GF <- countdata_mice[,col_S_CON_GF]
```
``` {r subset-check, include = FALSE}
# Check
colnames(counts_S_CON_GF)
# Remove variable no longer needed
rm(col_S_CON_GF)
```

Format Count and Annotation Data
```{r column-annot}
# DESeq2 requires a dataframe of column information for creating the DESeq Data Set
group_data <- str_remove(colnames(counts_S_CON_GF), pattern = ".S[0-9]+")
group_data <- factor(group_data)
coldata_mice <- data.frame(Group = group_data)
```

``` {r col-annot-check, include = FALSE}
group_data
table(group_data)
class(coldata_mice)
```

``` {r row-annot}
# EdgeR uses a dataframe to provide gene annotation information to the rows
# I'll use Entrez IDs and gene symbols
entrez_id_anot <- mapIds(org.Mm.eg.db, rownames(counts_S_CON_GF), keytype = "ENSEMBL", column = "ENTREZID")
symbols_anot <- mapIds(org.Mm.eg.db, rownames(counts_S_CON_GF), keytype = "ENSEMBL", column = "SYMBOL")
gene_mice <- data.frame(Entrez_ID = entrez_id_anot, Symbol = symbols_anot)
```

``` {r row-an-check, include = FALSE}
head(gene_mice, 3)
class(gene_mice)
# Remove unneeded variables
rm(entrez_id_anot, symbols_anot)
```

``` {r keep-symbol}
# The EdgeR tutorial by Chen et al. removed genes without an official symbol for the analysis. While there could be valid reason for keeping those genes, such as novel exploratory research, I will follow their example because it will keep the data size manageable and the annotations consistent for the purposes of this assignment. Further, the results spreadsheet provided by Stilling et al. for their analysis of this dataset did not have any of their reported genes with missing symbol information.
keep_genes <- !is.na(gene_mice$Symbol)
counts_S_CON_GF <- counts_S_CON_GF[keep_genes,]
gene_mice <- gene_mice[keep_genes,]
```

``` {r keep-check, include = FALSE}
sum(is.na(gene_mice$Symbol))
nrow(counts_S_CON_GF)
nrow(gene_mice)
```

Create & Explore Data Objects for EdgeR and DESeq2 Packages
``` {r edgeR-object}
# The count matrix and corresponding data is put together
ER_Counts <- DGEList(counts_S_CON_GF, group = group_data, genes = gene_mice)
# A few simple checks
class(ER_Counts)
# Shows group & library sizes
head(ER_Counts$samples, 4)
```

``` {r ER-check, include = FALSE}
# Explore what other information the data object includes
head(ER_Counts$counts, 4)
head(ER_Counts$genes, 4)
```

``` {r ER-des-m}
# In EdgeR, a design matrix is also used for some functions and can be created here. The design is specified simply in this case as ~0+group_data, so that samples are simply marked as part of a group or not.
ER_des.m <- model.matrix(~0+group_data)
colnames(ER_des.m) <- levels(group_data)
```

``` {r des-check, include = FALSE}
ER_des.m
```

```{r DESeq2-object}
# The count matrix and corresponding data is put together. The DESeq2 object also allows gene information to be added like in EdgeR, but under the argument, rowData
DS_Counts <- DESeqDataSetFromMatrix(countData = counts_S_CON_GF, colData = coldata_mice, rowData = gene_mice, design = ~ 0 + Group)
# A few simple checks
DS_Counts
#Shows library sizes
colSums(counts(DS_Counts))
```

``` {r DS-check, include = FALSE}
head(assay(DS_Counts), 4)
head(rowData(DS_Counts), 4)
```

Filtering and Quality Control
``` {r filters}
# The EdgeR has a function designed for filtering according to user-set parameters, which returns a logical vector that can be passed to subset the count data object.
# The default parameters for are min.count=10 (minimum numeric count for at least some samples), min.total.count = 15, and min.prop=0.7, with values being calculated as counts per million (CPM)
ER_filt_data_def <- filterByExpr(ER_Counts, ER_des.m)
table(ER_filt_data_def)
# Keeps 13855

# The DESeq2 tutorial filters out genes that have 1 or fewer reads total. The authors describe this only as a pre-filtering step to eliminate empty and nearly empty row, with further filtering occurring downstream.
DS_filt_data_def <- rowSums(counts(DS_Counts)) > 1
table(DS_filt_data_def)
# Keeps 18435

# Since the Stilling et al. paper that is the source of the data set did not specify their filtering parameters, but did mention using DESeq2 with default parameters for their analysis, I will filter both sets using the total count greater than 1.
# EdgeR has an internal library sizes variable, so we use the argument keep.lib.sizes = FALSE to reevaluate the library sizes after eliminating some genes.
# DESeq2 uses the sum of count data, and does not need the argument.
ER_Counts <- ER_Counts[DS_filt_data_def, , keep.lib.sizes=FALSE]
DS_Counts <- DS_Counts[DS_filt_data_def, ]
```

Library Normalization
```{r normalization}
# Library size normalization is where the two packages diverge significantly, each using a different statistical method
# In EdgeR, normalization is done by trimmed mean of M (TMM)
ER_Counts_N <- calcNormFactors(ER_Counts)
ER_Counts_N$samples

# DESeq2 uses "median ratio method" to estimate size factors
DS_Counts_N <- estimateSizeFactors(DS_Counts)
DS_Counts_N$sizeFactor

```

### Data Visualization

## 4. Main Software Tools Description

From De Paepe: TMM (EdgeR) gives a correction factor for the library size, but the median ratio method (DESeq2) generates a correction factor to apply to the raw counts.


## 5. Code Section 2 – Main Analysis

From the assignment details:

25%

"What would go in this section would depend upon your project, e.g.: writing code to build and test a classifier, build visualizations to explore ideas, conduct a statistical test of a biological hypothesis, or explore the impact of methodological choices upon results"

"Your main analysis section must also include a minimum of 2 figures and a maximum of 3 figures."

"Please see the guidelines for code section #1 above. The same general messages apply, including to focus upon quality, not length, and to comment your code well, etc."


##6. Quality of Visualizations

From assignment details:

20%

"Throughout, ensure that your figures are clear and well labeled. Even for simple figures, such as histograms, ensure that you have accurate, informative axis labels. Also, consider readability, visual appeal, and accessibility. Use well-differentiated colours, and avoid relying upon the red- green spectrum to convey scientifically important information. Remember, you can consider using a combination of colour and symbol/pattern to convey your meaning. The grade in this section is based upon quality and novelty, not having the maximum permissible number of figures. You should have a total of 4-6 figures for your project (excluding the bonus section, should you choose to complete that section)."


## 7. Bonus Section (Optional)

From the assignment details:
"An example of a topic that could go here would be benchmarking your analysis for computational speed and/or for accuracy (against comparator tools). Or, you could include supplementary statistical testing (e.g. testing whether your results are sensitive to your analytical choices in your main analysis section). You could also include supplementary visualizations of your choosing."

"The maximum length of this section is 2 pages of your final PDF assignment. Any content above the two-page limit for this section will not be graded. This could consist of prose, commented code, statistical results, and/or visualizations (your choice)."


## 8. Results and Discussion

From the assignment details:
"short written section of 2-3 paragraphs: 10%"
"Paragraph 1: Return to your original question. What is the answer to your question? What did you discover? Were your results as expected or not?"
"Paragraph 2: Briefly describe any key caveats of your study. For example, are the conclusions that can be drawn limited by sample size or any other concerns? Were there biases in data availability that could have impacted your project?"
"Paragraph 3: What would be the next steps for this research? What would you do next if you had more time and if you were going to develop this work into a larger project? Did your results reveal any interesting preliminary findings that would be worthy of follow-up study?"
"I suggest that you consider citing 2-4 references in your discussion section to help with
interpretation of your results. You may cite references that you also cited in your introduction"
"At the end, I also encourage you to add an additional short paragraph reflecting upon the process of completing this assignment. What did you learn through completing your project? What lessons will you take forward in your future coursework and career? Reflection can help us to solidify our learning and help us to act upon what we have learned in the future. This additional, optional paragraph is not included in the 3-paragraph length limit for this section."

## 9. Acknowledgements

Add any necessary acknowledgements here


## 10. References

From the assignment details:
" I suggest consulting and citing 3-6 papers from the literature to help you to develop your idea and/or interpret your results. Ten papers from the literature is the hard maximum for citations, reflecting that you have other assignments and commitments and to encourage some academics/life balance"

"Additional citations for vignettes, other tutorials, StackOverflow posts, etc., are not included in the 10-reference limit. You must cite all such sources."

"You would include scientific references as an in-text citation in the relevant sentence of your assignment, i.e. introduction or discussion (e.g. Xu et al. 2020). Also, list the full reference here at the end. If you used any specific online tutorials or a specific StackOverflow posting, for example, you must also include those here."



